# High-throughout virus screening
This repository contains updated scripts and pipelines used for screening multiple NGS datasets for viruses, collating the data and visualizing them simultaneously. These scripts result from work related to "High-throughput screening reveals high diversity and widespread distribution of viruses in black soldier flies (Hermetia illucens)" DOI: https://doi.org/10.1016/j.jip.2025.108322 . The original versions can also found at https://doi.org/10.5281/zenodo.14289390 and https://doi.org/10.5281/zenodo.13975607


Currently, this repository hosts two different approaches in screening pipelines. All future updates of either pipeline will be hosted here.

The piplines (baring the R scripts) are compatible for Linux only.

## First Approach - PoolingScreen

This pipeline utilises a retrograde BLASTx protocol on assembled contigs and transcripts using DIAMOND. Following this, the viral contigs from all of the datasets included in the pipeline run are collated together in a single file, which then is then analysed using CheckV. The main outputs to be utilised by the end user are the Krona charts at each stage and the CheckV results. The Krona charts can be viewed for each individual dataset, as well as combined charts by project and eventually for all of the datasets together.

While PoolingScreen was written as a virus screening pipeline, it can be used to screen for which ever taxa is preferred and included in the databases.

At the moment, PoolingScreen is available for Illumina single- and paired-end sequence datasets (e.g. MiSeq, Novaseq, etc.). There are plans to also adjust the pipeline for datasets of other sequencing technologies
    
More information on PoolingScreen can be found in the "PoolingScreen" directory.

![image](https://github.com/user-attachments/assets/08ee1cc7-2ce6-435d-88b8-e0d6e9317c57)

Visual diagram of the PoolingScreen pipeline.
Credit: Fig. 1 in Pienaar et al. 2025, DOI (https://doi.org/10.1016/j.jip.2025.108322)


## Second approach - Virus screening by high-throughput mapping and visualisation  

This pipeline is not fully automated and is segmented into three parts, 1) high-throughput read mapping and filtering for multiple samples and reference sequences, 2) visualisation of reference sequence annotations, and 3) high-throughput visualisation of BAM files as a heatmap.

Currently, the annotation visualisation output requires some post editing (in Inkscape/Photoshop) after being generated by the R script. This will be resolved for simple ORF annotations of viral sequences.
There are plans to automated both R visualisation scripts to work together in future updates.

More information can be found in the "HT_Mapping_pipeline" directory.


